### PR TITLE
Add ICU message format support

### PR DESCRIPTION
**Pull request by akozich**
Currently, the plugin uses java MessageFormat to find message parameters and generate methods accepting parameters for such messages.

This PR adds ICU message format support. [ICU message format](https://unicode-org.github.io/icu/userguide/format_parse/messages/) extends java MessageFormat adding support for new placeholder types including:
- plural
- select
- selectordinal

The most important part here from my perspective is plural which implements language specific [Plural Rules](https://cldr.unicode.org/index/cldr-spec/plural-rules)

[Examples](https://support.crowdin.com/icu-message-syntax/)

For conf file

```
en {
    choice = "{n,choice,-1#is negative| 0#is zero or fraction | 1#is one |1.0<is 1+ |2#is two |2<is more than 2.}"
    plural = "{dogsCount, plural, one {One dog is} other {# dogs are}}"
    selectordinal = "{year, selectordinal, one {#st} two {#nd} few {#rd} other {#th}}"
    select = "{gender, select, male {He uses} female {She uses} other {They use}}"
    multi = "{n,number,currency} {year, selectordinal, one {#st} two {#nd} three {#rd} other {#th}} {0}"
}
```

it will generate

```
package org.example.i18n

object Bundle {
  val languages: Map[String, I18N] = Map(("en", en))
  abstract class I18N {
    def choice(n: Long): String
    def multi(n: Double, year: Long, x0: Any): String
    def plural(dogsCount: Long): String
    sealed trait Gender
    object Gender {
      case object male extends Gender
      case object female extends Gender
      case object other extends Gender
    }
    def select(gender: Gender): String
    def selectordinal(year: Long): String
  }
  object en extends I18N {
    def choice(n: Long): String = com.ibm.icu.text.MessageFormat.format("""{n,choice,-1#is negative| 0#is zero or fraction | 1#is one |1.0<is 1+ |2#is two |2<is more than 2.}""", java.util.Map.of("n", n))
    def multi(n: Double, year: Long, x0: Any): String = com.ibm.icu.text.MessageFormat.format("""{n,number,currency} {year, selectordinal, one {#st} two {#nd} three {#rd} other {#th}} {0}""", java.util.Map.of("n", n, "year", year, "0", x0))
    def plural(dogsCount: Long): String = com.ibm.icu.text.MessageFormat.format("""{dogsCount, plural, one {One dog is} other {# dogs are}}""", java.util.Map.of("dogsCount", dogsCount))
    def select(gender: Gender): String = com.ibm.icu.text.MessageFormat.format("""{gender, select, male {He uses} female {She uses} other {They use}}""", java.util.Map.of("gender", gender))
    def selectordinal(year: Long): String = com.ibm.icu.text.MessageFormat.format("""{year, selectordinal, one {#st} two {#nd} few {#rd} other {#th}}""", java.util.Map.of("year", year))
  }
}
```

### CODE REVIEW

**Comment by ant8e** [comment]
Thanks for the PR.  Will review it as soon as I get the change to find some time (busy week)

File: src/main/scala/tech/ant8e/sbt/i18n/SbtI18nPlugin.scala [hunk 1]
```diff
@@ -39,9 +39,12 @@ object SbtI18nPlugin extends AutoPlugin {
             i18nBundlePackageName.value,
             i18nBreakOnMissingKeys.value
           ),
-          mappings in packageSrc ++= managedSources.value pair (Path
+          packageSrc / mappings ++= managedSources.value pair (Path
             .relativeTo(sourceManaged.value) | Path.flat),
-          sourceGenerators += generateI18NBundleTask.taskValue
+          sourceGenerators += generateI18NBundleTask.taskValue,
+          libraryDependencies ++= Seq(
```
Comment (akozich) [inline src/main/scala/tech/ant8e/sbt/i18n/SbtI18nPlugin.scala:45]: Not sure if this part will work. I'd like this dependency to be added to the project if the plugin is added.

File: project/build.properties [hunk 2]
```diff
@@ -1 +1 @@
-sbt.version=1.6.1
+sbt.version=1.10.1
```
Comment (akozich) [inline project/build.properties:1]: Had to bump SBT version to work with JDK 21

File: src/main/scala/tech/ant8e/sbt/i18n/BundleEmitter.scala [hunk 3]
```diff
@@ -42,31 +53,47 @@ case class BundleEmitter(config: Config, packageName: String, breakOnMissingKeys
 
   private def toScalaType(paramType: ParamType) =
     paramType match {
-      case StringParam => "String"
-      case DateParam   => "java.util.Date"
-      case DoubleParam => "Double"
-      case LongParam   => "Long "
+      case AnyParam     => "Any"
```
Comment (akozich) [inline src/main/scala/tech/ant8e/sbt/i18n/BundleEmitter.scala:56]: Changed `StringType` to `AnyType` as `{0}` pattern will be transformed to the regular toString invocation.

File: src/main/scala/tech/ant8e/sbt/i18n/BundleEmitter.scala [hunk 4]
```diff
@@ -42,31 +53,47 @@ case class BundleEmitter(config: Config, packageName: String, breakOnMissingKeys
...
+      case DateParam    => "java.util.Date"
+      case DoubleParam  => "Double"
+      case LongParam    => "Long"
+      case e: EnumParam => e.typeName
```
Comment (akozich) [inline src/main/scala/tech/ant8e/sbt/i18n/BundleEmitter.scala:60]: Special type for [Select syntax](https://support.crowdin.com/icu-message-syntax/#select). Values are exposed as ADT

File: src/main/scala/tech/ant8e/sbt/i18n/SbtI18nPlugin.scala [hunk 5]
```diff
@@ -39,11 +39,17 @@ object SbtI18nPlugin extends AutoPlugin {
...
+          packageSrc / mappings ++= managedSources.value pair (Path
             .relativeTo(sourceManaged.value) | Path.flat),
           sourceGenerators += generateI18NBundleTask.taskValue
         )
-    ) ++ Seq(i18nBundlePackageName := "org.example.i18n", i18nBreakOnMissingKeys := false)
+    ) ++ Seq(
+      i18nBundlePackageName  := "org.example.i18n",
+      i18nBreakOnMissingKeys := false,
+      libraryDependencies ++= Seq(
+        "com.ibm.icu" % "icu4j" % "75.1"
```
Comment (akozich) [inline src/main/scala/tech/ant8e/sbt/i18n/SbtI18nPlugin.scala:50]: It adds compile time dependency

**Comment by akozich** [comment]
Updated readme. There's one breaking change: methods generated for `date` or `time` parameters will accept `java.time.ZonedDateTime` instead of `Date`. Where should I add this?

**Comment by ant8e** [comment]
Now that I think of it, the best place to document the breaking changes would the release notes.  I will take care of it.

Now that I think of it, I'm a bit sad about the fact that the generated code will be only JVM compatible because of the icu4j dependency. I am not sure if there are actually Scala.js users of this plugin, but still. 
Do you think @akozich that it would be feasible to have a pure Scala alternative for this dependency?

**Comment by akozich** [comment]
1. There're thousands of lines of code in `icu4j`. I would not even try to rewrite it in pure scala 
2. On the other hand there was `java.text.MessageFormat` used. This plugin wasn't already js compatible

**Comment by ant8e** [comment]
Yes, you are right, I thought  `java.text.MessageFormat` was Scala.js friendly, but it is indeed not. 

So let's merge this !

**Pull request merged**